tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java

```diff
@@ -0,0 +1,256 @@
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.ClusterResource;
+import org.apache.kafka.common.ClusterResourceListener;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.ObsoleteBrokerException;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.utils.Utils;
+
+import java.util.List;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.CountDownLatch;
+
+public class CompatibilityTest {
+    public static class CompatibilityTestDeserializer implements Deserializer<byte[]>, ClusterResourceListener {
+        static volatile boolean expectClusterId = false;
+
+        @Override
+        public void configure(Map<String, ?> configs, boolean isKey) {
+            // nothing to do
+        }
+
+        @Override
+        public byte[] deserialize(String topic, byte[] data) {
+            return data;
+        }
+
+        @Override
+        public void close() {
+            // nothing to do
+        }
+
+        @Override
+        public void onUpdate(ClusterResource clusterResource) {
+            if (expectClusterId) {
+                if (clusterResource.clusterId() == null) {
+                    throw new RuntimeException("expected cluster id to be " +
+                            "supported, but it was null.");
+                }
+            } else {
+                if (clusterResource.clusterId() != null) {
+                    throw new RuntimeException("expected cluster id to be " +
+                            "null, but it was supported.");
+                }
+            }
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+        ArgumentParser parser = argParser();
+        Namespace res = null;
+        try {
+            res = parser.parseArgs(args);
+        } catch (ArgumentParserException e) {
+            if (args.length == 0) {
+                parser.printHelp();
+                System.exit(0);
+            } else {
+                parser.handleError(e);
+                System.exit(1);
+            }
+        }
+        String bootstrapServer = res.getString("bootstrap_server");
+        String topic = res.getString("topic");
+        boolean offsetsForTimesSupported = res.getBoolean("offsets_for_times_supported");
+        CompatibilityTestDeserializer.expectClusterId =
+                res.getBoolean("cluster_id_supported");
+
+        Properties producerProps = new Properties();
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG,
+                "org.apache.kafka.common.serialization.ByteArraySerializer");
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
+                "org.apache.kafka.common.serialization.ByteArraySerializer");
```

Comment:
Instead of setting it this way, you can just pass the serializers to the constructor.